engine/src/main/java/io/zeebe/engine/nwe/behavior/BpmnEventSubscriptionBehavior.java

```diff
@@ -0,0 +1,199 @@
+
+  public void publishTriggeredBoundaryEvent(final BpmnElementContext context) {
+
+    elementInstanceState.getDeferredRecords(context.getElementInstanceKey()).stream()
+        .filter(record -> record.getValue().getBpmnElementType() == BpmnElementType.BOUNDARY_EVENT)
+        .findFirst()
+        .ifPresent(
+            deferredRecord ->
+                activateBoundaryEvent(context, deferredRecord.getKey(), deferredRecord.getValue()));
+  }
+
+  private void activateBoundaryEvent(
+      final BpmnElementContext context,
+      final long elementInstanceKey,
+      final WorkflowInstanceRecord eventRecord) {
+
+    streamWriter.appendNewEvent(
+        elementInstanceKey, WorkflowInstanceIntent.ELEMENT_ACTIVATING, eventRecord);
+
+    stateBehavior.createElementInstanceInFlowScope(context, elementInstanceKey, eventRecord);
+    stateBehavior.spawnToken(context);
+  }
+
+  public <T extends ExecutableCatchEventSupplier> Either<Failure, Void> subscribeToEvents(
+      final T element, final BpmnElementContext context) {
+
+    try {
+      catchEventBehavior.subscribeToEvents(context.toStepContext(), element);
+      return Either.right(null);
+
+    } catch (final MessageCorrelationKeyException e) {
+      return Either.left(
+          new Failure(
+              e.getMessage(),
+              ErrorType.EXTRACT_VALUE_ERROR,
+              e.getContext().getVariablesScopeKey()));
+
+    } catch (final EvaluationException e) {
+      return Either.left(
+          new Failure(
+              e.getMessage(), ErrorType.EXTRACT_VALUE_ERROR, context.getElementInstanceKey()));
+    }
+  }
+
+  public void unsubscribeFromEvents(final BpmnElementContext context) {
+    catchEventBehavior.unsubscribeFromEvents(
+        context.getElementInstanceKey(), context.toStepContext());
+  }
+
+  public void publishTriggeredEventSubProcess(final BpmnElementContext context) {
+    final var elementInstance = stateBehavior.getElementInstance(context);
+
+    if (isInterrupted(elementInstance)) {
+      elementInstanceState.getDeferredRecords(context.getElementInstanceKey()).stream()
+          .filter(record -> record.getKey() == elementInstance.getInterruptingEventKey())
+          .filter(record -> record.getValue().getBpmnElementType() == BpmnElementType.SUB_PROCESS)
+          .findFirst()
+          .ifPresent(
+              record -> {
+                final var elementInstanceKey = record.getKey();
+                final var interruptingRecord = record.getValue();
+
+                streamWriter.appendNewEvent(
+                    elementInstanceKey,
+                    WorkflowInstanceIntent.ELEMENT_ACTIVATING,
+                    interruptingRecord);
+
+                stateBehavior.createChildElementInstance(
+                    context, elementInstanceKey, interruptingRecord);
+              });
+    }
+  }
+
+  private boolean isInterrupted(final ElementInstance elementInstance) {
+    return elementInstance.getNumberOfActiveTokens() == 2
```

Comment:
what is this magic number? :sweat_smile:

Reply:
Yeah. This is scary :scream: 
This is from the current implementation. We will tackle it when looking at the state.